### PR TITLE
tikv-server: use clap to parse command line options

### PR DESCRIPTION
**Pull request by andelf**
tikv-server -h

```
TiKV 0.0.1
PingCAP Inc. <info@pingcap.com>
A Distributed transactional key-value database powered by Rust and Raft

USAGE:
    tikv-server [OPTIONS] --config <FILE>

FLAGS:
    -h, --help       Prints help information
    -V, --version    Prints version information

OPTIONS:
    -A, --addr <IP:PORT>              Sets listening address [default: 127.0.0.1:20160]
        --advertise-addr <IP:PORT>    Sets advertise listening address for client communication [default: ${addr}]
        --capacity <CAPACITY>         Sets the store capacity [default: 0]
    -C, --config <FILE>               Sets config file
        --data-dir <PATH>             Sets the path to store directory [default: /tmp/tikv/store]
        --labels <KEY=VALUE>...       Sets server labels [default: ]
    -f, --log-file <FILE>             Sets log file
    -L, --log-level <LEVEL>           Sets log level [values: trace, debug, info, warn, error, off]
        --pd-endpoints <PD_URL>...    Sets PD endpoints [default: 127.0.0.1:2379]

```

### CODE REVIEW

**Comment by andelf** [comment]
for comparation:

```
Usage: ./bin/tikv-server [options]

Options:
    -A, --addr default is 127.0.0.1:20160
                        set listening address
        --advertise-addr if not set, use ${addr} instead.
                        set advertise listening address for client
                        communication
    -L, --log log level: trace, debug, info, warn, error, off
                        set log level
    -f, --log-file if not set, output log to stdout
                        set log file
    -V, --version       print version information
    -h, --help          print this help menu
    -C, --config file path
                        set configuration file
    -s, --store /tmp/tikv/store
                        set the path to store directory
        --capacity default: 0 (disk capacity)
                        set the store capacity
        --pd 127.0.0.1:2379,127.0.0.1:3379
                        pd endpoints
        --labels zone=example-zone,disk=example-disk
                        attributes about this server
```

**Comment by andelf** [comment]
BREACKING CHANGES:

now ``--config FILE`` is a must.

**Comment by andelf** [comment]
DNM: bug in handling default options

**Comment by siddontang** [comment]
do we not support "--pd" flag?  I just see "--pd-endpoints"

**Comment by andelf** [comment]
``   .aliases(&["pd", "[pd-endpoint"])``

**Comment by siddontang** [comment]
why must we need config file now? I suggest keeping compatibility.

**Comment by siddontang** [comment]
Any update?  @andelf

**Comment by andelf** [comment]
It seems no one use tikv without a config file.

**Comment by siddontang** [comment]
In our previous design, the config file is for advanced use, it is not requisite.

It is fine for a user to just try TiKV easily.

File: src/bin/tikv-server.rs [hunk 1]
```diff
@@ -855,74 +856,99 @@ fn run_raft_server(pd_client: RpcClient,
...
+            .help("Sets listening address"))
+        .arg(Arg::with_name("advertise-addr")
+            .long("advertise-addr")
+            .takes_value(true)
+            .value_name("IP:PORT")
+            .help("Sets advertise listening address for client communication"))
+        .arg(Arg::with_name("log-level")
+            .short("L")
+            .long("log-level")
+            .alias("log")
```
Comment (siddontang) [inline src/bin/tikv-server.rs]: do we need this alias?

**Comment by andelf** [comment]
ok, I'll fix this.

**Comment by siddontang** [comment]
PTAL @overvenus @hhkbp2

File: src/bin/tikv-server.rs (hunk 1 above)
Comment (andelf) [inline src/bin/tikv-server.rs]: will not display in `help`. no harm

**Comment by hhkbp2** [comment]
LGTM

**Comment by siddontang** [comment]
CI failed @andelf

**Comment by siddontang** [comment]
Any update? @andelf

**Comment by andelf** [comment]
It seems that rustfmt's result is not stable?
ref https://travis-ci.org/pingcap/tikv/jobs/230249926 
I've add a (git diff) while format check fails.

**Comment by disksing** [comment]
LGTM.

**Comment by siddontang** [comment]
Does CircleCI not use own binary rustfmt?

**Comment by siddontang** [comment]
You can fix the CI now @andelf

File: src/raft/log_unstable.rs [hunk 2]
```diff
@@ -339,30 +332,44 @@ mod test {
...
-           (vec![new_entry(5, 1), new_entry(6, 1), new_entry(7, 1)], 5, None,
-               vec![new_entry(6, 2)],
-                   5, vec![new_entry(5, 1), new_entry(6, 2)]),
-
-            (vec![new_entry(5, 1), new_entry(6, 1), new_entry(7, 1)], 5, None,
-                vec![new_entry(7, 2), new_entry(8, 2)],
-                    5, vec![new_entry(5, 1), new_entry(6, 1), new_entry(7, 2), new_entry(8, 2)]),
-      ];
+        let tests =
+            vec![// replace to the end
```
Comment (BusyJay) [inline src/raft/log_unstable.rs]: I think the original form is more readable.

File: src/raft/raft_log.rs [hunk 3]
```diff
@@ -447,25 +447,28 @@ mod test {
...
-                new_entry(5, 4)], 4),
-            (vec![new_entry(2, 2), new_entry(3, 3), new_entry(4, 4), new_entry(5, 4)], 4),
-            (vec![new_entry(3, 3), new_entry(4, 4), new_entry(5, 4)], 4),
-            (vec![new_entry(4, 4), new_entry(5, 4)], 4),
-            // conflicts with existing entries
-            (vec![new_entry(1, 4), new_entry(2, 4)], 1),
-            (vec![new_entry(2, 1), new_entry(3, 4), new_entry(4, 4)], 2),
-            (vec![new_entry(3, 1), new_entry(4, 2), new_entry(5, 4), new_entry(6, 4)], 3),
-        ];
+        let tests =
```
Comment (BusyJay) [inline src/raft/raft_log.rs]: These seems unrelated to this pr.

File: src/raft/raft_log.rs [hunk 4]
```diff
@@ -879,48 +887,103 @@ mod test {
...
-             Some(last_index - 2), last_index - 2, true), // conflict with existing committed entry
-            (last_term - 2, last_index - 2, last_index, vec![new_entry(last_index - 1, 4),
-                                                             new_entry(last_index, 4)],
-             Some(last_index), last_index, false),
-        ];
-
-        for (i, &(log_term, index, committed, ref ents, wlasti, wcommit, wpanic)) in tests.iter()
-            .enumerate() {
+        let tests =
+            vec![// not match: term is different
```
Comment (BusyJay) [inline src/raft/raft_log.rs]: This is really unreadable and too long.

File: src/bin/tikv-server.rs [hunk 5]
```diff
@@ -19,10 +19,11 @@
 #![allow(never_loop)]
 #![allow(needless_pass_by_value)]
 
+#[macro_use]
```
Comment (BusyJay) [inline src/bin/tikv-server.rs]: Why use macro?

Comment (andelf) [inline src/bin/tikv-server.rs]: clap use macro to fetch crate name & version

File: src/raft/raft_log.rs (hunk 4 above)
Comment (andelf) [inline src/raft/raft_log.rs]: ok, I'll revert changes.

**Comment by andelf** [comment]
restart travis-ci job to pass test
